### ISSUE TITLE
Rendering html without cloning the whole page

### ISSUE THREAD

**Issue by eng1neer** [niklasvh/html2canvas#1284]
Is there a way to render an element without cloning the whole page? I noticed that it takes 70-80% of the whole processing time to clone the document.

**Comment by niklasvh** [comment]
Not currently possible

**Comment by mmahlies** [comment]
what happened to this features?

**Comment by mojimi** [comment]
@niklasvh This is specially slow since it will reload all script/css tags, please look into this.

Why does it even need to clone the window? I don't understand.

**Comment by Sonamkhurana** [comment]
Have you found any solution to make this process fast?

**Comment by K-D-K** [comment]
@eng1neer As for now I have a hack to reduce processing time.
Let us assume I need to create a canvas for div with id="html2canvas_canvas_div"
I have added a class named as "**html2canvas_whitelist_class**"  to all it parents up to **html** tag element and to all the children and children's children recursively.

whitelisted those elements using [ignoreElements](https://html2canvas.hertzen.com/configuration/) option

By above-mentioned hack, style elements won't get cloned to the destination document.
For that, I have added the required styles in single style sheets and copied those style sheets to the cloned document.
It steeply increases the performance

`{
	getStyleSheets: function() {
		var styleSheets = this.data.styleSheets;
		if (!styleSheets) {
			var fileNames = ["required_css_file.css"];
			styleSheets = document.styleSheets.filter(function(sheet) {
				var fileLocation = sheet.href;
				if (fileLocation) {
					return !fileNames.every(function(fileName) {
						return !fileLocation.endsWith(fileName);
					})
				}
			})
			this.data.styleSheets = styleSheets;
		}
		return styleSheets;
	},
	copyStyles: function(destDocument) {
		var styleElement = destDocument.createElement("style");
		destDocument.body.appendChild(styleElement);
		var styleElementSheet = styleElement.sheet;

		this.getStyleSheets().forEach(function(styleSheet) {
			styleSheet.rules.forEach(function(rule) {
				styleElementSheet.insertRule(rule.cssText);
			})
		})
	}
}`

**Comment by fds-github** [comment]
I don't think the gist actually works as is and is just meant to be an outline of an approach, right?  It doesn't define isElementToSkip, it seems to rely on a global "selector", and maybe some more stuff as I go further.  Or am I missing something?

If there's not a better way, I'll use this as a starting point.

**Comment by K-D-K** [comment]
> I don't think the gist actually works as is and is just meant to be an outline of an approach, right? It doesn't define isElementToSkip, it seems to rely on a global "selector", and maybe some more stuff as I go further. Or am I missing something?
> 
> If there's not a better way, I'll use this as a starting point.

For my scenario I need to skip few elements which should not be part of final image.
I know the selectors of those images.
Added the **isElementToSkip**  method in the gist
You can check now.

**Comment by fds-github** [comment]
I was still getting errors, such as accessing `this.data.styleSheets` and calling `filter()` on an object instead of an array.  I've [hacked it up a bit more](https://gist.github.com/fds-github/51c362f904f8e10fe6c973fd558a1998) to fix these problems and I think it's now easier to use:  As long as you're not passing options to html2canvas, you should be able to replace calls like `html2canvas(element)` with
`html2canvasPruned.html2canvas(element)`.

For me, this cuts the time to render the canvas to about 10% of what it used to be.  A great improvement!  Obviously, the savings depend on your DOM structure.

I'm having problems trying to make multiple images in parallel.  With vanilla html2canvas this works fine for me, but with this change, images get truncated.  Originally I thought it might be a problem with the clean-up I do in removeWhiteListClass, but removing that doesn't solve the problem.

**Comment by fds-github** [comment]
I serialized all of my calls and the problems remain for some images when I use this pruning, but not if I don't.  My guess is that something in my CSS renders things differently if siblings are missing, but I haven't been able to track it down yet.

**Comment by K-D-K** [comment]
Hi @fds-github
`this.data.styleSheets` here I store all information in `this.data` object...
This weekend will try to check the "export of multiple images in parallel" issue.
If possible can you share the sample code/project to test/reproduce the issue?

**Comment by fds-github** [comment]
Thanks for keeping this on your radar.  Sorry I haven't sent anything:  Pretty swamped at the moment.  Reading my previous comment now and trying to remember the details, I think it was far from clear that there were problems exporting multiple images in parallel, since I was seeing similar problems when I serialized the calls to remove the parallelism.  Instead, I suspect that some of my CSS selectors might be rendering differently with the pruned tree.  I think the changes I included in my 11.11.21 comment might be helpful to people, but otherwise I'd guess that the problem is on my side.
